### PR TITLE
W.I.P: Zend: container/offset semantics RFC: object handlers

### PR DESCRIPTION
**Pull request by Girgias**
Conversions:
 - [x] ext/dom: All classes
 - [x] ext/simplexml: SimpleXML
 - [x] ext/pdo: PDORow
 - [x] ext/intl: ResourceBundle
 - [x] Zend: WeakMap
 - [x] ext/ffi FFI classes
 - [x] ext/spl: ArrayObject
 - [x] ext/spl: ArrayIterator
 - [x] ext/spl: RecursiveArrayIterator
 - [x] ext/spl: SplObjectStorage
 - [x] ext/spl: SplFixedArray

Classes that implement `ArrayAccess` without specialized handlers
 - [ ] ext/phar: Phar
 - [ ] ext/phar: PharData
 - [ ] ext/spl: CachingIterator

Other To-dos:
 - [x] Fix Windows build
 - [x] JIT Support
 - [ ] opcache support (opcache_persist stuff)
 - [x] Fix issue not being able to call offsetFetch() as it is not marked to return by reference?
 - [ ] Be sure about the indirect thing for ArrayObject, and possibly use it for SplFixedArray
 - [ ] RFC Updating

### CODE REVIEW

File: Zend/zend_dimension_handlers.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+//BEGIN_EXTERN_C()
+
+#include "zend_types.h"
+
+typedef struct _zend_class_dimensions_functions {
+	/* rv is a slot provided by the callee that is returned */
+	zval *(*/* const */ read_dimension)(zend_object *object, zval *offset, zval *rv);
+	bool  (*/* const */ has_dimension)(zend_object *object, zval *offset);
+	zval *(*/* const */ fetch_dimension)(zend_object *object, zval *offset, zval *rv);
```
Comment (morrisonlevi) [inline Zend/zend_dimension_handlers.h:31]: I don't understand why `rv` is passed, and something is also returned. Just pass in `rv`, and make it return void, or if you need a success/failure determination, then return `zend_result` instead? Is there something I am missing? Some case where the `rv` isn't the exact same thing as what is returned, other than perhaps on error?

Same comment goes for `fetch_append` and `read_dimension`, I suppose.

Comment (morrisonlevi) [inline Zend/zend_dimension_handlers.h:31]: I'm not understanding something. Why do we pass in a `zval *rv`, which I assume is the "return value", but then also return it? Wouldn't it be better to pass in the return value, and return `void` or `zend_result` instead? Are they separate zvals in some cases other than an error?

File: Zend/zend_interfaces_dimensions.stub.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+interface DimensionUnsetable
+{
+    public function offsetUnset(mixed $offset): void;
+}
+
+interface Appendable
+{
+    public function append(mixed $value): void;
```
Comment (morrisonlevi) [inline Zend/zend_interfaces_dimensions.stub.php:47]: Can we please consider `offsetAppend` instead, and have the relevant SPL objects proxy their `append` to `offsetAppend`? As I pointed out to you in Stack Overflow chat, there are real-world projects with `append` methods that don't match this signature nor semantics, which will make upgrading more painful (but not a BC break, just a road block to onboarding the new APIs). Probably changes the interface to `DimensionAppendable` as well.

File: Zend/zend_dimension_handlers.h (hunk 1 above)
Comment (Girgias) [reply Zend/zend_dimension_handlers.h:31]: Yes, there are seperate zvals in some cases, in which case we need to make the returned zval IS_INDIRECT. The lact of *rv as an arg and return value has been an issue with properly supporting fetches for the property hook RFC as the get_ptr_ptr handler doesn't have this mechanism.

File: Zend/zend_interfaces_dimensions.stub.php (hunk 2 above)
Comment (Girgias) [reply Zend/zend_interfaces_dimensions.stub.php:47]: Please provide this feedback on the internals mailing list.

**Comment by Girgias** [comment]
See PR on the main repo: https://github.com/php/php-src/pull/14920

**Pull request closed without merging**